File: Harvest/Models/UserProfile.swift
import Foundation

struct UserProfile: Codable, Identifiable, Sendable {
    let id: String
    var email: String
    var nickname: String?
    var age: Int?
    var bio: String?
    var location: String?
    var gender: String?
    var preferences: String?
    var goals: String?
    var hobbies: [String]?
    var photos: [String]?
    var distancePreference: Int?
    var interestedIn: [String]?
    var lookingFor: String?
    var heightCm: Int?
    var smoking: String?
    var drinking: String?
    var cannabis: String?
    var spiritualOrientation: String?
    var childrenStatus: String?
    var onboardingCompleted: Bool?
    var createdAt: String?
    var updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case id, email, nickname, age, bio, location, gender, preferences, goals, hobbies, photos
        case distancePreference = "distance_preference"
        case interestedIn = "interested_in"
        case lookingFor = "looking_for"
        case heightCm = "height_cm"
        case smoking
        case drinking
        case cannabis
        case spiritualOrientation = "spiritual_orientation"
        case childrenStatus = "children_status"
        case onboardingCompleted = "onboarding_completed"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    var displayName: String {
        nickname ?? email.components(separatedBy: "@").first ?? "User"
    }

    var primaryPhoto: String? {
        photos?.first
    }

    var goalsList: [String] {
        guard let goals else { return [] }

        let trimmedGoals = goals.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmedGoals.isEmpty { return [] }

        if trimmedGoals.hasPrefix("["),
           let data = trimmedGoals.data(using: .utf8),
           let decoded = try? JSONDecoder().decode([String].self, from: data) {
            return decoded
                .map { Self.normalizeGoalLabel($0.trimmingCharacters(in: .whitespacesAndNewlines)) }
                .filter { !$0.isEmpty }
        }

        if trimmedGoals.contains(",") {
            return trimmedGoals
                .components(separatedBy: ",")
                .map { Self.normalizeGoalLabel($0.trimmingCharacters(in: CharacterSet(charactersIn: "\"[] ").union(.whitespacesAndNewlines))) }
                .filter { !$0.isEmpty }
        }

        let normalizedGoal = Self.normalizeGoalLabel(
            trimmedGoals.trimmingCharacters(in: CharacterSet(charactersIn: "\"[] ").union(.whitespacesAndNewlines))
        )
        return normalizedGoal.isEmpty ? [] : [normalizedGoal]
    }

    var lifestyleDetails: [(label: String, value: String)] {
        var details: [(label: String, value: String)] = []

        if let lookingFor = lookingForDisplayValue {
            details.append(("Looking For", lookingFor))
        }
        if let height = heightDisplayValue {
            details.append(("Height", height))
        }
        if let smoking = smokingDisplayValue {
            details.append(("Smoking", smoking))
        }
        if let drinking = drinkingDisplayValue {
            details.append(("Drinking", drinking))
        }
        if let cannabis = cannabisDisplayValue {
            details.append(("Cannabis", cannabis))
        }
        if let spiritualOrientation = spiritualOrientationDisplayValue {
            details.append(("Spiritual Orientation", spiritualOrientation))
        }
        if let childrenStatus = childrenStatusDisplayValue {
            details.append(("Children", childrenStatus))
        }

        return details
    }

    private var lookingForDisplayValue: String? {
        guard let lookingFor else { return nil }
        let normalized = Self.normalizeGoalLabel(lookingFor.trimmingCharacters(in: .whitespacesAndNewlines))
        return normalized.isEmpty ? nil : normalized
    }

    private var heightDisplayValue: String? {
        guard let heightCm else { return nil }
        return HeightFormatter.string(from: heightCm)
    }

    private var smokingDisplayValue: String? {
        Self.formatLifestyleValue(smoking)
    }

    private var drinkingDisplayValue: String? {
        Self.formatLifestyleValue(drinking)
    }

    private var cannabisDisplayValue: String? {
        Self.formatLifestyleValue(cannabis)
    }

    private var spiritualOrientationDisplayValue: String? {
        Self.formatLifestyleValue(spiritualOrientation)
    }

    private var childrenStatusDisplayValue: String? {
        Self.formatLifestyleValue(childrenStatus)
    }

    private static func normalizeGoalLabel(_ value: String) -> String {
        switch value.lowercased() {
        case "short-term dating", "casual":
            return "Dating"
        case "long-term relationship", "long-term commitment", "long_term_commitment":
            return "Long-term Commitment"
        case "relationship":
            return "Relationship"
        case "marriage":
            return "Marriage"
        case "not sure yet", "not sure":
            return "Dating"
        default:
            return value
        }
    }

    private static func formatLifestyleValue(_ value: String?) -> String? {
        guard let value else { return nil }

        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        return trimmed
            .replacingOccurrences(of: "_", with: " ")
            .split(separator: " ")
            .map { word in
                let lower = word.lowercased()
                if ["and", "to", "not"].contains(lower) {
                    return lower
                }
                return word.prefix(1).uppercased() + word.dropFirst().lowercased()
            }
            .joined(separator: " ")
    }
}

File: android/app/src/main/java/com/harvestglass/harvest/data/model/UserProfile.kt
package com.harvestglass.harvest.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import com.harvestglass.harvest.util.HeightFormatter

/** Mirrors Harvest/Models/UserProfile.swift. */
@Serializable
data class UserProfile(
    val id: String,
    val email: String? = null,
    val nickname: String? = null,
    val age: Int? = null,
    val bio: String? = null,
    val location: String? = null,
    val gender: String? = null,
    val preferences: String? = null,
    val goals: String? = null,
    val hobbies: List<String>? = null,
    val photos: List<String>? = null,

    @SerialName("distance_preference") val distancePreference: Int? = null,
    @SerialName("interested_in") val interestedIn: List<String>? = null,
    @SerialName("looking_for") val lookingFor: String? = null,
    @SerialName("height_cm") val heightCm: Int? = null,
    val smoking: String? = null,
    val drinking: String? = null,
    val cannabis: String? = null,
    @SerialName("spiritual_orientation") val spiritualOrientation: String? = null,
    @SerialName("children_status") val childrenStatus: String? = null,
    @SerialName("relationship_status") val relationshipStatus: String? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,

    @SerialName("onboarding_completed") val onboardingCompleted: Boolean? = null,
    @SerialName("is_banned") val isBanned: Boolean? = null,
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("updated_at") val updatedAt: String? = null,

    // Values / Soil tab display settings.
    @SerialName("values_blurb") val valuesBlurb: String? = null,
    @SerialName("show_values_brought") val showValuesBrought: Boolean? = null,
    @SerialName("show_values_sought") val showValuesSought: Boolean? = null,
    @SerialName("show_values_blurb") val showValuesBlurb: Boolean? = null,
    @SerialName("show_values_graph") val showValuesGraph: Boolean? = null,
    /** "need" | "bring" — which side the profile radar shows. */
    @SerialName("profile_graph_side") val profileGraphSide: String? = null,

    // Notification preferences.
    @SerialName("notif_messages_enabled") val notifMessagesEnabled: Boolean? = null,
    @SerialName("notif_matches_enabled") val notifMatchesEnabled: Boolean? = null,
    @SerialName("notif_likes_enabled") val notifLikesEnabled: Boolean? = null,
    @SerialName("notif_gardener_local_enabled") val notifGardenerLocalEnabled: Boolean? = null,
    @SerialName("notif_gardener_local_hour") val notifGardenerLocalHour: Int? = null
) {
    val displayName: String
        get() = nickname ?: email?.substringBefore('@')?.takeIf { it.isNotEmpty() } ?: "User"

    val primaryPhoto: String? get() = photos?.firstOrNull()

    /**
     * Goals have been stored three ways over the app's life — a JSON array, a
     * comma-joined string (what onboarding writes), and a bare single value —
     * so all three are accepted and normalised to today's labels.
     */
    val goalsList: List<String>
        get() {
            val trimmed = goals?.trim().orEmpty()
            if (trimmed.isEmpty()) return emptyList()

            if (trimmed.startsWith("[")) {
                runCatching { Json.decodeFromString<List<String>>(trimmed) }
                    .getOrNull()
                    ?.let { decoded ->
                        return decoded.map { normalizeGoalLabel(it.trim()) }.filter { it.isNotEmpty() }
                    }
            }

            if (trimmed.contains(",")) {
                return trimmed.split(",")
                    .map { normalizeGoalLabel(it.trim(*GOAL_TRIM_CHARS)) }
                    .filter { it.isNotEmpty() }
            }

            return normalizeGoalLabel(trimmed.trim(*GOAL_TRIM_CHARS))
                .takeIf { it.isNotEmpty() }
                ?.let { listOf(it) }
                ?: emptyList()
        }

    val heightDisplayValue: String? get() = heightCm?.let { HeightFormatter.string(it) }

    /** Label/value pairs for the profile card, in the order iOS shows them. */
    val lifestyleDetails: List<Pair<String, String>>
        get() = buildList {
            heightDisplayValue?.let { add("Height" to it) }
            lookingFor?.trim()?.let { normalizeGoalLabel(it) }?.takeIf { it.isNotEmpty() }
                ?.let { add("Looking For" to it) }
            formatLifestyleValue(smoking)?.let { add("Smoking" to it) }
            formatLifestyleValue(drinking)?.let { add("Drinking" to it) }
            formatLifestyleValue(cannabis)?.let { add("Cannabis" to it) }
            formatLifestyleValue(spiritualOrientation)?.let { add("Spiritual Orientation" to it) }
            formatLifestyleValue(childrenStatus)?.let { add("Children" to it) }
        }

    private companion object {
        val GOAL_TRIM_CHARS = charArrayOf('"', '[', ']', ' ', '\n', '\t')

        /** Words that stay lowercase when title-casing a stored value. */
        val SMALL_WORDS = setOf("and", "to", "not")

        fun normalizeGoalLabel(value: String): String = when (value.lowercase()) {
            "short-term dating", "casual" -> "Dating"
            "long-term relationship", "long-term commitment", "long_term_commitment" ->
                "Long-term Commitment"
            "relationship" -> "Relationship"
            "marriage" -> "Marriage"
            "not sure yet", "not sure" -> "Dating"
            else -> value
        }

        fun formatLifestyleValue(value: String?): String? {
            val trimmed = value?.trim().orEmpty()
            if (trimmed.isEmpty()) return null

            return trimmed
                .replace("_", " ")
                .split(" ")
                .filter { it.isNotEmpty() }
                .joinToString(" ") { word ->
                    val lower = word.lowercase()
                    if (lower in SMALL_WORDS) lower
                    else word.take(1).uppercase() + word.drop(1).lowercase()
                }
        }
    }
}
